domain/application/state/application.go

```diff
@@ -2193,6 +2194,412 @@ FROM v_revision_updater_application_unit
+// configuration.
+func (st *State) SetApplicationConfig(ctx context.Context, appID coreapplication.ID, config map[string]application.ApplicationConfig) error {
+	db, err := st.DB()
+	if err != nil {
+		return internalerrors.Capture(err)
+	}
+
+	ident := applicationID{ID: appID}
+
+	// We don't need to verify the application exists on write, as there
+	// will be a foreign key constraint that will fail if the application
+	// doesn't exist.
+	getQuery := `
+SELECT &applicationConfig.*
+FROM v_application_config
+WHERE uuid = $applicationID.uuid;
+`
+	deleteQuery := `
+DELETE FROM application_config
+WHERE application_uuid = $applicationID.uuid
+AND key IN ($S[:]);
+`
+	insertQuery := `
+INSERT INTO application_config (*)
+VALUES ($setApplicationConfig.*);
+`
+	updateQuery := `
+UPDATE application_config
+SET value = $setApplicationConfig.value,
+	type_id = $setApplicationConfig.type_id
+WHERE application_uuid = $setApplicationConfig.application_uuid;
+`
+	settingsQuery := `
+INSERT INTO application_setting (*)
+VALUES ($setApplicationSettings.*)
+ON CONFLICT(application_uuid) DO UPDATE SET
+	trust = excluded.trust;
+`
+
+	getStmt, err := st.Prepare(getQuery, applicationConfig{}, ident)
+	if err != nil {
+		return internalerrors.Errorf("preparing query for application config: %w", err)
+	}
+	deleteStmt, err := st.Prepare(deleteQuery, ident, sqlair.S{})
+	if err != nil {
+		return internalerrors.Errorf("preparing query for application config: %w", err)
+	}
+	insertStmt, err := st.Prepare(insertQuery, setApplicationConfig{})
+	if err != nil {
+		return internalerrors.Errorf("preparing query for application config: %w", err)
+	}
+	updateStmt, err := st.Prepare(updateQuery, setApplicationConfig{})
+	if err != nil {
+		return internalerrors.Errorf("preparing query for application config: %w", err)
+	}
+	settingsStmt, err := st.Prepare(settingsQuery, setApplicationSettings{})
+	if err != nil {
+		return internalerrors.Errorf("preparing query for application config: %w", err)
+	}
+
+	trust, err := encodeTrustConfig(config)
+	if err != nil {
+		return internalerrors.Errorf("encoding trust config: %w", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var current []applicationConfig
+		if err := tx.Query(ctx, getStmt, ident).GetAll(&current); err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return internalerrors.Errorf("querying application config: %w", err)
+		}
+
+		currentM := make(map[string]applicationConfig)
+		for _, c := range current {
+			currentM[c.Key] = c
+		}
+
+		// Work out what we need to do, based on what we have, vs what we
+		// need.
+		var removals sqlair.S
+		var updates []setApplicationConfig
+		for k := range currentM {
+			if _, ok := config[k]; !ok {
+				removals = append(removals, k)
+				continue
+			}
+
+			typeID, err := encodeConfigType(config[k].Type)
+			if err != nil {
+				return internalerrors.Errorf("encoding config type: %w", err)
+			}
```

Comment:
Maybe?
```
if config[k].Value == current[k].Value {
    continue
}
```

(or whatever the correct syntax would be)

Reply:
I'll check if we're not racing here, but if not, I think that's a fine optimization. :+1:

Reply:
We also need to ensure the type is the same, but this is now done.